tinygrad/renderer/cstyle.py

```diff
@@ -338,14 +338,19 @@ def render_kernel(self, function_name, kernel, bufs, uops, prefix=None):
     for dtype in dedup(uop.dtype for uop in uops if uop.dtype in {dtypes.half, dtypes.bfloat16}):
       prefix += [f"#include <cuda_{'fp' if dtype == dtypes.half else 'bf'}16.h>"] + [self.render_vector_prefix(dtype.vec(sz)) for sz in [4, 8]]
 
-    # TODO: this has to be way better to generate for arbitrary M,N,K: use arg[1] for MNK, use arg[4] for vec sizes, encode register packing
-    dt_map = { dtypes.float: "f32", dtypes.half: "f16", dtypes.bfloat16: "bf16" }
-    for arg in dedup([uop.arg for uop in uops if uop.op is UOps.WMMA]):
-      fn, ti, to, ci, co = arg[0], self.render_dtype(arg[2]), self.render_dtype(arg[3]), dt_map[arg[2]], dt_map[arg[3]]
-      prefix.append(f"""__device__ {to}4 __{fn}({ti}8 a, {ti}4 b, {to}4 c) {{ int *a_pk = (int *) (&a), *b_pk = (int *) (&b);
-asm( "mma.sync.aligned.m16n8k16.row.col.{co}.{ci}.{ci}.{co} {{ %0, %1, %2, %3 }}, {{ %4, %5, %6, %7 }}, {{ %8, %9 }}, {{ %0, %1, %2, %3 }};"
-  : "+f"(c.x), "+f"(c.y), "+f"(c.z), "+f"(c.w) : "r"(a_pk[0]), "r"(a_pk[1]), "r"(a_pk[2]),  "r"(a_pk[3]), "r"(b_pk[0]), "r"(b_pk[1]) );
-return c;}}""")
+    dt_map = { dtypes.half: "f16", dtypes.bfloat16: "bf16" }
+    for name, (N, M, K), dtype_in, dtype_out, _, _, upcast_axes, _ in dedup([uop.arg for uop in uops if uop.op is UOps.WMMA]):
+      sza, szb, szc = (prod(sz for _, sz in upc) for upc in upcast_axes)
+      dtype_a, dtype_b, dtype_c = (self.render_dtype(dt.vec(sz)) for dt,sz in ((dtype_in,sza),(dtype_in,szb),(dtype_out,szc)))
+      operands_c = [f"%{c}" for c in range(szc*dtype_out.itemsize//4)] # %0, %1, ... operands
+      operands_a = [f"%{a+len(operands_c)}" for a in range(sza*dtype_in.itemsize//4)]
+      operands_b = [f"%{b+len(operands_c+operands_a)}" for b in range(szb*dtype_in.itemsize//4)]
+      c_pks = ", ".join([f'"+f"(c.{_nms[i]})' for i in range(szc*dtype_out.itemsize//4)]) # "+f"(c.x), "+f"(c.y), ...
+      a_pks, b_pks = (", ".join([f'"r"({v}_pk[{i}])' for i in range(sz*dtype_in.itemsize//4)]) for sz,v in ((sza,'a'),(szb,'b'))) # "r"(a_pk[0]), ...
+      prefix.append(f"""__device__ {dtype_c} __{name}({dtype_a} a, {dtype_b} b, {dtype_c} c){{\n  int *a_pk = (int *)(&a), *b_pk = (int *)(&b);
```

Comment:
Why is f32 hardcoded now? What about f16 output?

Reply:
All current tensor core supported for CUDA have float as accumulator and both Turing and tf32 accumulate with float as well

Reply:
I removed it from the map since it’s not needed. All accumulate on f32 as previously mentioned, and when support for additional tensor core shapes is added, two maps will be required. This is because, for tf32, when a CUDA tensor core receives f32 as input, it actually processes it as tf32 but outputs it in f32. Therefore, we will need a mapping like `dtype_in_map={dtypes.float: "tf32"}` and `dtype_out_map={dtypes.float: "f32"}`.